### PR TITLE
feat: list arguments for resource actions in class diagrams

### LINKED ISSUE #392: Show arguments (maybe only required arguments) in generated class diagrams
At the moment, for something like

```elixir
create :register do
  argument :email, :ci_string do
    allow_nil? false
  end
  
  argument :password, :ci_string do
     allow_nil? false
  end
  
  argument :confirm_password, :ci_string do
    allow_nil? false
  end
end
```

We would only see something like

```
register()
```

in the generated diagram. Would be great to show arguments there.

### PR DESCRIPTION
**Pull request by AliNisarAhmed**
This PR resolves #392 by printing arguments and attributes for resource actions in class diagrams, with an upper limit of 4 (the rest hidden with ellipsis)

# Contributor checklist

- [ ] Bug fixes include regression tests
- [ ] Chores
- [ ] Documentation changes
- [x] Features include unit/acceptance tests
- [ ] Refactoring
- [ ] Update dependencies

### CODE REVIEW

File: lib/ash/api/diagram/diagram.ex [hunk 1]
```diff
@@ -123,6 +123,15 @@ defmodule Ash.Api.Info.Diagram do
     Enum.map_join(list, "\n", fn item -> "#{indent}#{indent}#{template_fn.(item)}" end)
   end
 
+  defp list_arguments([]), do: ""
+
+  defp list_arguments(arguments) do
+    arguments
+    |> Enum.map(&"#{class_short_type(&1.type)} #{&1.name}")
```
Comment (joshprice) [inline lib/ash/api/diagram/diagram.ex:130]: You'll want to use `Enum.map_join` like above. Credo whinges about this as I discovered.

**Review by joshprice — COMMENTED** [review]
Looking good so far

File: lib/ash/api/diagram/diagram.ex (hunk 1 above)
Comment (AliNisarAhmed) [reply lib/ash/api/diagram/diagram.ex]: Fixed, thanks

**Comment by zachdaniel** [comment]
After the arguments, we'll also want to list the accepted attributes (for non-read actions). Something like:

```elixir
action.accept
|> Enum.map(&Ash.Resource.Info.attribute(resource, &1))
# [%Attribute{type: type, name: name, allow_nil?: true/false}, ...]
```

**Comment by AliNisarAhmed** [comment]
@zachdaniel @joshprice Let me know from below diagrams which one you prefer. 

I am limiting the shown arguments to 4, and the rest either turn to ellipsis, or I show their count.

Personally I prefer `+ n more` style, with default values. We could provide options for users to enable/disable printing of types and default args.

Also, I could not find a tool-tip option for Mermaid class diagrams. I think it is only an option for Flowcharts.


- Ellipsis: 

ellipsis

- Ellipsis with default arguments: 

ellipsis_with_default_args

- `+ n` more style 

plus_more

- `+ n` more with default args: 

plus_more_with_default

**Comment by joshprice** [comment]
I like the ellipsis stylistically but also think the extra info is nice too. What about both with `UUID x_id, ...+2 more)`?

**Comment by AliNisarAhmed** [comment]
@zachdaniel @joshprice The diagram below prints both arguments and attributes, and limits arguments to max of 4, and shows the count of rest of the args, omitting the default values to reduce noise. Let me know how it looks now

image

**Comment by zachdaniel** [comment]
I think we should do a *total limit* of 4, for both arguments and attributes, and they should all go before the ellipsis.

**Comment by zachdaniel** [comment]
That looks excellent!

**Comment by AliNisarAhmed** [comment]
@zachdaniel I would have added tests, but I think the current class diagram tests cover all cases, showcasing no arguments, 4 arguments, and 4+ arguments, so I think they suffice, let me know. 

If things look fine I will squash the commits and open the PR for review.

**Comment by zachdaniel** [comment]
I agree :) Lets get it merged 🚀

**Comment by AliNisarAhmed** [comment]
@zachdaniel ready for review

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Pull request merged**